packages/playwright-core/src/server/socksClientCertificatesInterceptor.ts

```diff
@@ -21,10 +21,49 @@ import fs from 'fs';
 import tls from 'tls';
 import stream from 'stream';
 import { createSocket } from '../utils/happy-eyeballs';
-import { globToRegex, isUnderTest } from '../utils';
+import { globToRegex, isUnderTest, ManualPromise } from '../utils';
 import type { SocksSocketClosedPayload, SocksSocketDataPayload, SocksSocketRequestedPayload } from '../common/socksProxy';
 import { SocksProxy } from '../common/socksProxy';
 import type * as channels from '@protocol/channels';
+import { debugLogger } from '../utils/debugLogger';
+
+class ALPNCache {
+  private _cache = new Map<string, ManualPromise<string>>();
```

Comment:
Q: Should we clear the cache at some point? Maybe make it bound to the `SocksProxyConnection` instance?

Reply:
Yeah, one per proxy instance sounds better.